### ISSUE TITLE
Regent Isomorphic Partitions

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#184]
It should be possible for two regions in Regent to share the same index partition.  This means that if you make two regions with the same index space, and you partition one of them, then you have to be able to get the corresponding partition for the other region.  Here is a made-up example that you can modify syntactically as you see fit (I mainly care about functionality):

var is = ispace(ptr,1024)
var x = region(is, double)
var y = region(is, double)
var x_part = partition(equal, x, ispace(int1d,4))
var y_part = partition(y, x_part.color) -- This line is made up
regentlib.assert(y_part.color == x_part.color, "should be true")

**Comment by elliottslaughter** [comment]
I have another idea for doing this that I've planning for a long time but haven't gotten around to.

Regent should understand the dual relationship between index spaces and regions. It should be possible to partition an index space and have that automatically apply to the region tree.

I'd imagine the code looking something like:

```
var is = ispace(ptr,1024)
var x = region(is, double)
var y = region(is, double)
var is_part = partition(equal, is, ispace(int1d,4)) -- Partition the ispace
var x_part = x[is_part] -- Turn the ispace partition into a region partition
var x_sub0 = x_part[0] -- Get a subregion like normal
var y_sub0 = y[is_part[0]] -- Or do the same thing by grabbing a sub-ispace and turning that into a region
regentlib.assert(y_sub0.color == x_sub0.color, "should be true")
```

This requires a change to the constraint checking in the type system to teach it about constraints on index spaces and the relationship between index spaces and regions. There would be two new edges 'is-region-of-ispace' and 'is-ispace-of-region' (possibly with nicer names) to help navigate around the graph; constraints would be stored purely on the ispace graph.

**Comment by lightsighter** [comment]
Yup, I was intentionally trying to propose something simpler than that to avoid perturbing your current implementation of the type system too much, but I agree that actually doing the partitioning on the index spaces would make the most sense since it mirrors how the runtime does it.

**Comment by streichler** [comment]
@elliottslaughter what's the current thinking on this one?

**Comment by elliottslaughter** [comment]
I think it's still important to do this, it's just low priority since we've got a workaround that's good enough for most of our users. But we did this in the previous version of Regent and I don't see why we can't do it again.
